_posts/2020-09-30-sparklyr-1.4.0-released/sparklyr-1.4.0.Rmd

```diff
@@ -0,0 +1,368 @@
+
+It is often observed that many machine learning algorithms perform better on numeric inputs that are standardized. Many of us have learned in stats 101 that given a random variable $X$, we can compute its mean $\mu = E[X]$, standard deviation $\sigma = \sqrt{E[X^2] - (E[X])^2}$, and then obtain a standard score $z = \frac{X - \mu}{\sigma}$ which has mean of 0 and standard deviation of 1.
+
+However, notice both $E[X]$ and $E[X^2]$ from above are quantities that can be easily skewed by extreme outliers in $X$, causing distortions in $z$.
+
+An alternative way of standardizing $X$ based on its median, 1st quartile, and 3rd quartile values, all of which are robust against outliers, would be the following:
+
+$\displaystyle z = \frac{X - \text{Median}(X)}{\text{P75}(X) - \text{P25}(X)}$
+
+and this is precisely what [RobustScaler](https://spark.apache.org/docs/3.0.0/api/java/org/apache/spark/ml/feature/RobustScaler.html) offers.
+
+To see `ft_robust_scaler()` in action and demonstrate its usefulness, we can go through a contrived example consisting of the following steps:
+
+- Draw 100 random samples from the standard normal distribution
+
+```{r echo=TRUE, eval=FALSE}
+sample_values <- rnorm(100)
+print(sample_values)
+```
+
+```
+  [1] -0.885440337  0.831855863 -0.890427366  0.437780168  0.091902152
+  [6]  0.774895164 -0.839350781 -1.173025189 -0.442174680 -0.002573251
+  ...
+```
+- Inspect the minimal and maximal values among the 100 random samples:
+
+```{r echo=TRUE, eval=FALSE}
+print(min(sample_values))
+```
+```
+  [1] -2.377851
+```
+
+```{r echo=TRUE, eval=FALSE}
+print(max(sample_values))
+```
+```
+  [1] 2.629998
+```
+
+- Now create 5 other values that are extreme outliers compared to the 100 random samples above. Given that we know all 100 samples are within the range of $(-3, 3)$, we can choose $21, 22, \ldots, 25$ as our outliers:
+
+```{r echo=TRUE, eval=FALSE}
+outliers <- 20 + seq(5)
+```
+
+- Copy all 105 values into a Spark dataframe named `sdf`
+
+```{r echo=TRUE, eval=FALSE}
+library(sparklyr)
+
+sc <- spark_connect(master = "local", version = "3.0.0")
+sdf <- copy_to(sc, data.frame(value = c(sample_values, outliers)))
+```
+
+- We can then apply `ft_robust_scaler()` to obtain the standardized value for each input:
+
+```{r echo=TRUE, eval=FALSE}
+scaled <- sdf %>%
+  ft_vector_assembler("value", "input") %>%
+  ft_robust_scaler("input", "scaled") %>%
+  dplyr::pull(scaled) %>%
+  unlist()
+```
+
+- Plotting the result shows the non-outlier data points are scaled to values that still more or less form a bell-shaped distribution, as expected, so the scaling is robust against influences of the outliers:
+
+```{r echo=TRUE, eval=FALSE}
+library(ggplot2)
+
+ggplot(data.frame(scaled = scaled), aes(x = scaled)) + geom_histogram(binwidth = 0.2)
+```
+
+![](images/scaled.png){#id .class width=40% height=40%}
```

Comment:
I think these plots would look better if we make them wider and use the light theme. I think the theme from the BBC looks quite nice, see https://bbc.github.io/rcookbook/ -- The book uses a similar theme but the code is a bit messy, see https://github.com/r-spark/the-r-in-spark/blob/master/r/plots.R

That said, this is minor feedback only worth addressing if you feel like it.